busctl: Add introspect support for methods with same name but different signature

AsamK opened the pull request:
D-Bus interfaces can have multiple methods with the same name, as long as they have different arguments (signature).
Currently busctl can call those methods but when introspecting the interface it just displays "Duplicate method"

This PR fixes the behavior, by also adding the signature to the hash for the members set.

**Before this patch:**
```sh
$ busctl --user introspect org.asamk.Signal /org/asamk/Signal 
Invalid introspection data: duplicate method 'sendMessage' on interface 'org.asamk.Signal'.
```

**After this patch:**

```sh
$ busctl --user introspect org.asamk.Signal /org/asamk/Signal 
NAME                                TYPE      SIGNATURE RESULT/VALUE FLAGS
org.asamk.Signal                    interface -         -            -
.sendMessage                        method    as        x            -
.sendMessage                        method    s         x            -
```

Calling the methods already works as expected, as the user must specify the signature explicitely:
```sh
busctl --user call org.asamk.Signal /org/asamk/Signal org.asamk.Signal sendMessage "as" 2 foo bar
busctl --user call org.asamk.Signal /org/asamk/Signal org.asamk.Signal sendMessage "s" foo
```
**XML introspection data:**
`$ busctl --user --xml introspect org.asamk.Signal /org/asamk/Signal`
```xml
<!DOCTYPE node PUBLIC "-//freedesktop//DTD D-BUS Object Introspection 1.0//EN" "http://www.freedesktop.org/standards/dbus/1.0/introspect.dtd">
<node name="/org/asamk/Signal">
 <interface name="org.asamk.Signal">
  <method name="sendMessage" >
   <arg type="as" direction="in"/>
   <arg type="x" direction="out"/>
  </method>
  <method name="sendMessage" >
   <arg type="s" direction="in"/>
   <arg type="x" direction="out"/>
  </method>
 <interface name="org.freedesktop.DBus.Introspectable">
  <method name="Introspect">
   <arg type="s" direction="out"/>
  </method>
 </interface>
 <interface name="org.freedesktop.DBus.Peer">
  <method name="Ping">
  </method>
 </interface>
</node>
```

poettering commented:
Ah um. Overloading methods is a very questionnable thing. Note sure if the D-Bus spec ever mentions this explicitly, but this is a very fishy concept, given that this is typically not universally supportable in all languages. @smcv what's your take on this?

smcv commented:
> D-Bus interfaces can have multiple methods with the same name, as long as they have different arguments (signature)

D-Bus implementations typically have a low-level message-oriented layer and build higher-level APIs (if present) on top of it, and the wire protocol is defined in terms of low-level message-oriented actions, which means there are lots of things that are technically possible to do but not interoperable. This is one of them.

The interoperable subset is not particularly well-defined: one day it would be nice if the D-Bus Specification documented what it was. However, having two methods of the same name in the same interface is certainly not in the interoperable subset. In many higher-level D-Bus implementations/bindings, calling such methods will require bypassing the higher-level (object-oriented) layer and using lower-level (message-oriented) APIs. 

If these methods appear in introspection XML, many implementations' code-generation tools will get confused, crash, or output non-compilable code. For example, I suspect gdbus-codegen will emit C code that can't be compiled, because 1-argument `Foo()` and 2-argument `Foo()` will both come out as `my_interface_call_foo()`.

busctl's introspection and tab-completion support are not exactly a high-level API, but they fall into the same general category - a convenience layer that works well as long as everyone sticks to the interoperable subset of D-Bus, but will get do the wrong thing or get confused if you go beyond the interoperable subset.

I would recommend naming your methods differently.

(If you're renaming them to disambiguate the overloaded names, also take the opportunity to use the conventional case combination: D-Bus method, signal and property names are traditionally all `CapitalizedWords`.)

AsamK commented:
Thanks for your input!
yep, gdbus-codegen generates non-compilable code for that interface, but other high-level APIs, like the java implementation, can handle it.
My dbus service api can definitely be improved, also in regard to other dbus conventions.

Apart from that, the question remains if busctl should be able to introspect such interfaces.
The dbus specification doesn't forbid it and busctl already supports calling the overloaded methods, so I don't see a reason why it shouldn't also allow introspecting interfaces with such methods.

poettering commented:
The way I see it, busctl should be debugging tool, and thus also a validating tool, and refuse invalid data. Hence I think we should refuse the data, but output more useful, debuggable log messages about it I think.

Introspection is ultimately just decoration. I think if we can argue that we can call functions that are not uniquely defined then that actually makes things work for people. But showing introspection data like that just hides errors I think, and it doesn't really make anything work that matters anyway, since it's only decoration...

Or to say this differently: the benefit of clearly pinpointing problems in your service outweighs the benefit of being able to show you the introspection data that is problematic.

I will thus prepare a patch that improves the error message, but I am very sure we shouldn't merge this patch

poettering commented:
I posted a PR improving the log messages now in #15843. Let's close this PR. I hope that's OK.

keszybz commented:
@AsamK please add the description from the PR to the commit message (comments on the PR are not easy to look up from git history, but the commit messages are immutable and easy to find). Also, please add output from `busctl --xml introspect` for your service to that commit message, so that we can see how the details look.

AsamK commented:
I've updated the commit message.

I also noticed that the improved error message for duplicate methods didn't work, because the hashset doesn't return `-EEXIST` for already existing entries, but instead just returns `0`. So duplicate methods didn't trigger an error but were just ignored.
I fixed this in a second commit to this PR, but I could also create a separate PR for it, if you prefer.

keszybz commented:
Hmm, looking at this again, I think this should be merged.
The first commit is a clear bugfix, so the only question is the second one.
I think the low-level library should allow flexibility. Validation should happen in upper layers.
But we should at least re-run the CI, I'll restart it.

mrc0mmand commented:
> Hmm, looking at this again, I think this should be merged. The first commit is a clear bugfix, so the only question is the second one. I think the low-level library should allow flexibility. Validation should happen in upper layers. But we should at least re-run the CI, I'll restart it.

That's going to need a rebase, since the PR tree doesn't even contain the correct CI configs.

yuwata commented on src/busctl/busctl.c:
```diff
@@ -809,7 +809,7 @@ static int on_interface(const char *interface, uint64_t flags, void *userdata) {
         r = set_put(members, m);
-        if (r == -EEXIST)
+        if (r == 0)
                 return log_error_errno(r,  "Invalid introspection data: duplicate interface '%s'.", interface);
```
Please do not pass `0` to `log_error_errno()`, but use `SYNTHETIC_ERRNO()` macro.
```
return log_error_errno(SYNTHETIC_ERRNO(EEXIST), "...");
```
Please also drop doubled white space.

keszybz replied:
Thanks, I somehow missed this. Updated.

yuwata commented on src/busctl/busctl.c:
```diff
@@ -852,7 +852,7 @@ static int on_method(const char *interface, const char *name, const char *signat
         r = set_put(members, m);
-        if (r == -EEXIST)
+        if (r == 0)
                 return log_error_errno(r, "Invalid introspection data: duplicate method '%s' on interface '%s'.", name, interface);
```
Also here, and below.

yuwata commented:
The first commit is LGTM too. Maybe, we should cherry pick the first one?
